projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,154 @@
+    storage->channel_enable[channel] = false;
+    storage->channel_readings[channel] = 0;
+    storage->channel_callback[channel] = NULL;
+    storage->callback_context[channel] = NULL;
+  }
+
+  status_ok_or_return(prv_reset(storage));
+  status_ok_or_return(prv_setup_register(
+      storage, ADS1015_ADDRESS_POINTER_CONFIG, CONFIG_REGISTER_MSB(ADS1015_CHANNEL_0),
+      CONFIG_REGISTER_LSB));  // sets the starting channel to channel 0
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_init_pin(&storage->ready_pin, &gpio_settings));
+  return gpio_it_register_interrupt(&storage->ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                    prv_interrupt_handler, storage);
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+// It also registers the interrupt handler on ALRT/RDY pin.
+StatusCode ads1015_configure_channel(Ads1015Storage *storage, Ads1015Channel channel, bool enable,
+                                     Ads1015Callback callback, void *context) {
+  if (storage == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  // if no channels were already enabled, set to given channel
+  if (!storage->channel_enable[0] && !storage->channel_enable[1] && !storage->channel_enable[2] &&
+      !storage->channel_enable[3]) {
+    status_ok_or_return(prv_set_channel(storage, channel));
+  }
+  storage->channel_enable[channel] = enable;
+  storage->channel_callback[channel] = callback;
+  storage->callback_context[channel] = context;
+  return STATUS_CODE_OK;
+}
+
+// reads raw 12 bit conversion results which are expressed in two's complement format
+StatusCode ads1015_read_raw(Ads1015Storage *storage, Ads1015Channel channel, int16_t *reading) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL || reading == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  *reading = storage->channel_readings[channel] >> NUM_RESERVED_BITS_CONV_REG;
+  return STATUS_CODE_OK;
+}
+
+// reads conversion value in mVolt
+StatusCode ads1015_read_converted(Ads1015Storage *storage, Ads1015Channel channel,
+                                  int16_t *reading) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL || reading == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  int16_t raw_12bit;
```

Comment:
We usually prefer to initialize variables to an invalid default value before passing them into a function. That's so that we don't risk using an uninitialized value (imagine if you forgot a status check somewhere and didn't actually set `raw_12bit` before using it)